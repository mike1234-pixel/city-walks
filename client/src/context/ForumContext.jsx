import { createContext, useEffect, useState } from "react"
import { useHistory } from "react-router-dom"
import axios from "axios"
import BoardBox from "../components/pages/LoginPage/LoggedInView/Components/BoardBox/BoardBox"

export const ForumContext = createContext();

export const ForumContextProvider = (props) => {
 
    const [loadingBoards, setLoadingBoards] = useState(true)
    const [boards, setBoards] = useState([])
    const [selectedThreads, setSelectedThreads] = useState([])

    const [currentBoardName, setCurrentBoardName] = useState("")
    const [currentBoardId, setCurrentBoardId] = useState("")

    let displayBoards;

    useEffect(() => {
    axios
    .get('http://localhost:5000/boards')
    .then((res, err) => {
        if (err) {
            console.log(err)
            alert("Could not connect to the server. Please try again.")
        } else {
            console.log(res.data)
            setBoards(res.data)
            console.log(boards)
            setLoadingBoards(false)
        }

    })}, []);

    if (!loadingBoards) {
    displayBoards = boards.map((board, index) => {
        return (
            <BoardBox boardId={board._id} name={board.name} description={board.description} index={index} />
                )
        })
    }

    return (
        <ForumContext.Provider 
            value={{
                loadingBoards,
                displayBoards,
                boards, // data
                selectedThreads, 
                setSelectedThreads,
                currentBoardName,
                setCurrentBoardName,
                currentBoardId,
                setCurrentBoardId
            }}>
            {props.children}
        </ForumContext.Provider>
    )
}