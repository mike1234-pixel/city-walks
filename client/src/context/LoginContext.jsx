import { createContext, useState } from "react"
import axios from "axios"
import qs from "qs"

export const LoginContext = createContext();

export const LoginContextProvider = (props) => {
    const [firstName, setFirstName] = useState("")
    const [lastName, setLastName] = useState("")
    const [registrationEmail, setRegistrationEmail] = useState("")
    const [registrationPassword, setRegistrationPassword] = useState("")
  
    const [loginEmail, setLoginEmail] = useState("")
    const [loginPassword, setLoginPassword] = useState("")
  
    const [loggedIn, setLoggedIn] = useState(false)
    const [userFirstName, setUserFirstName] = useState("")
    const [userLastName, setUserLastName] = useState("")
  
    const [verificationEmail, setVerificationEmail] = useState("")

    const [forgotPasswordEmail, setForgotPasswordEmail] = useState("")

    const handleChangeRegistration = (event) => {
        switch(event.target.name) {
            case "registration-fname":
              setFirstName(event.target.value)
              break;
            case "registration-lname":
              setLastName(event.target.value)
              break;
            case "registration-email":
              setRegistrationEmail(event.target.value)
              break;
            case "registration-password":
              setRegistrationPassword(event.target.value)
              break;
          } 
    }
  
    const handleSubmitRegistration = (event) => {
      console.log("handle submit triggered")
        event.preventDefault()
  
        const payload = {
            fname: firstName,
            lname: lastName,
            email: registrationEmail,
            password: registrationPassword,
          };
  
    axios
        .post("http://localhost:5000/register-user", qs.stringify(payload))
        .then((res, err) => {
          if (err) {
            console.log(err);
          } else if (res.data === "An account with this email already exists.") {
            alert("An account with this email already exists.")
          } else if (res.data === "We have sent you an email. Please verify your account by clicking the link in the mail.") {
            alert("We have sent you an email. Please verify your account by clicking the link in the email. (This code expires after 10 minutes)")
           } else {
            alert("Registration Complete")
            setLoggedIn(true)
            setFirstName("")
            setLastName("")
            setRegistrationEmail("")
            setRegistrationPassword("")
            setUserFirstName(res.data.fname)
            setUserLastName(res.data.lname)
            window.scrollTo(0, 0)
          }
        });
    }
  
    const handleChangeLogin = (event) => {
      switch(event.target.name) {
          case "login-email":
            setLoginEmail(event.target.value)
            break;
          case "login-password":
            setLoginPassword(event.target.value)
            break;
        } 
  }
  
  const handleSubmitLogin = (event) => {
    console.log("handle submit login triggered")
      event.preventDefault()
  
      const payload = {
          email: loginEmail,
          password: loginPassword
        };
  
  axios
      .post("http://localhost:5000/login-user", qs.stringify(payload))
      .then((res, err) => {
        if (err) {
          console.log(err);
        } else if (res.data === "Your account exists but is not activated. Please click 'verify account' for email verification.") {
          alert("Your account exists but is not activated. Please click 'verify account' for email verification.")
        } else if (res.data === "unsuccessful login attempt") {
          alert("Unsuccessful Login Attempt. Please Try Again.")
        } else {
          alert("Login Complete")
          setLoggedIn(true)
          setLoginEmail("")
          setLoginPassword("")
          setUserFirstName(res.data.fname)
          setUserLastName(res.data.lname)
          window.scrollTo(0, 0)
        }
      });
  }

  // resend verification email

  const handleChangeVerification = (event) => {
    switch(event.target.name) {
        case "verification-email":
          setVerificationEmail(event.target.value)
          break;
      }   
  }

const handleSubmitVerification = (event) => {
    console.log("handle submit verification")
    event.preventDefault()

    const payload = {
        email: verificationEmail,
      };

axios
    .post("http://localhost:5000/reverify-user", qs.stringify(payload))
    .then((res, err) => {
      if (err) {
        console.log(err);
      } else {
        alert("Verification email sent. Check your inbox.")
        setVerificationEmail("")
        window.scrollTo(0, 0)
      }
    })
}

  // forgot password

  const handleChangeForgotPassword = (event) => {
    switch(event.target.name) {
        case "forgot-password-email":
          setForgotPasswordEmail(event.target.value)
          break;
      }   
  }

const handleSubmitForgotPassword = (event) => {
    console.log("handle submit forgot password")
    event.preventDefault()

    const payload = {
        email: forgotPasswordEmail,
      };

axios
    .post("http://localhost:5000/forgot-password", qs.stringify(payload))
    .then((res, err) => {
      if (err) {
        console.log(err);
      } else {
        alert("We have sent you an email. Please click the click in your email to reset your password")
        setForgotPasswordEmail("")
        window.scrollTo(0, 0)
      }
    })
}
  
  const logOut = () => {
    setLoggedIn(false)
    setUserFirstName("")
    setUserLastName("")
  }

    return (
        <LoginContext.Provider 
            value={{
                firstName: firstName,
                lastName: lastName,
                registrationEmail: registrationEmail,
                registrationPassword: registrationPassword,
                handleChangeRegistration: handleChangeRegistration,
                handleSubmitRegistration: handleSubmitRegistration,
                loggedIn: loggedIn,
                // login state and functions
                loginEmail: loginEmail,
                loginPassword: loginPassword,
                handleChangeLogin: handleChangeLogin,
                handleSubmitLogin: handleSubmitLogin,
                // logged in user data
                userFirstName: userFirstName,
                userLastName: userLastName,
                // resend verification email
                verificationEmail: verificationEmail,
                handleChangeVerification: handleChangeVerification,
                handleSubmitVerification: handleSubmitVerification,
                // forgot password
                forgotPasswordEmail: forgotPasswordEmail,
                handleChangeForgotPassword: handleChangeForgotPassword,
                handleSubmitForgotPassword: handleSubmitForgotPassword,
                // logout function
                logOut: logOut
            }}
        >
            {props.children}
        </LoginContext.Provider>
    )
}