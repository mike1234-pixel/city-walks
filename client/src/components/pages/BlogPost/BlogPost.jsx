import { useEffect, useContext, useState } from "react"
import { BlogsContext } from "../../../context/BlogsContext"
import { LoginContext } from "../../../context/LoginContext"
import toTitleCase from "../../../functions/toTitleCase"
import { MDBInput, MDBBtn, MDBIcon } from "mdbreact"
import axios from "axios"
import qs from "qs"
import './BlogPost.css'

const BlogPost = ({match}) => {

    // useEffect(() => {
    //     window.scrollTo(0, 0);
    //   });

      const blogTitle = toTitleCase(match.url.replace("/blog/", "").replace(/-/g, " "))

      const { blogPosts, blogsLoading } = useContext(BlogsContext)
      const { loggedIn, userFirstName, userId } = useContext(LoginContext)

      const [comment, setComment] = useState("")
      const [currentBlogTitle, setCurrentBlogTitle] = useState("")

      const handleChange = (event) => {
        setComment(event.target.value)
      }

      const handleSubmit = (event) => {
        event.preventDefault()

        let payload = {
          currentBlogTitle: blogTitle,
          comment: comment,
          userFirstName: userFirstName,
          userId: userId
        };

      axios
      .post("http://localhost:5000/add-blog-comment", qs.stringify(payload))
      .then((err) => {
        if (err) {
          console.log(err);
        }
      });

      alert("Comment Submitted")
      setComment("")
      window.scrollTo(0, 0);
      }

      let post = "loading"

      if (!blogsLoading) {

      let selectedBlogPost = blogPosts.filter((post) => post.title === blogTitle)
      selectedBlogPost = selectedBlogPost[0]

      if (selectedBlogPost === undefined) {
        post = "walk not found"
      } else {
        post = 
        <div>
            <h1>{selectedBlogPost.title}</h1>
            <p>{selectedBlogPost.content}</p>
            <img src={selectedBlogPost.img}/>
        </div>
      }
    }

    return (
        <div>
            <p>BLOG POST (redirect successful)</p>
            <p>{post}</p>
            {loggedIn &&
              <form onSubmit={handleSubmit} className="add-blog-comment-form">
                <MDBInput type="textarea" rows="5" name="comment" id="comment" value={comment} label="comment" onChange={handleChange} required/>
                <MDBBtn outline color="elegant" type="submit">
                    Add Comment <MDBIcon icon="plus"/>
                </MDBBtn>
              </form>
            }
        </div>
    )
}

export default BlogPost