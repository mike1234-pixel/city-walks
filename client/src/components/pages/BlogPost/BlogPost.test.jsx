import React from 'react'
import renderer from 'react-test-renderer'
import { BlogsContextProvider } from "../../../context/BlogsContext"
import BlogPost from './BlogPost'

// snapshot test
it('Blog Post renders correctly', () => {  

  const tree = renderer
    .create(
        <BlogsContextProvider>
            <BlogPost match={{params: {id: 1}, isExact: true, path: "", url: "/blog/blog-title-1"}}/>
        </BlogsContextProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});