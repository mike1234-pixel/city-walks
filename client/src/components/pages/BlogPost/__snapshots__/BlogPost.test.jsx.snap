// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog Post renders correctly 1`] = `
<div>
  <p>
    BLOG POST (redirect successful)
  </p>
  <p>
    loading
  </p>
</div>
`;
